[wasm] Webpack & NextJS & TypeScript samples

pavelsavara opened the pull request:
## Samples
- `console-node-ts` shows how the `dotnet.d.ts` allows compile time validation of the embedding API
- `browser-webpack` is sample which is bundling `dotnet.js` and `underscore` library into single `main.js`. It's minified in Release mode.
- `browser-nextjs` shows how to create react component and re-use runtime, when the component is instantiated multiple times

All 3 samples consume `dotnet.js` via `@microsoft/dotnet-runtime` in-tree **npm package**. The package is installed into local `node_modules` during managed build, so that Release/Debug configuration matches the C# side.

## Other changes
- changed npm package scope to `@microsoft/dotnet-runtime`
- set correct version to runtime package.json during build
- fixed type definition for `DotnetModuleConfig`
- updated `js_string_to_mono_string` to return `MonoStringNull` which is `0` instead of `null`
- updated `dotnet.d.ts` banner

pavelsavara committed:
component (+3 squashed commit)

Squashed commit:

[2f3e017331c] nextJs sample

[7d21d5e11cc] webpack sample

[7cf26e30f45] wip

kant2002 commented on src/mono/sample/wasm/browser-nextjs/package-lock.json:
```diff
@@ -0,0 +1,7418 @@
+  "packages": {
+    "": {
+      "dependencies": {
+        "dotnet": "file:../../../../../artifacts/bin/native/net7.0-Browser-Debug-wasm/",
```
Is there a way how to use configuration dependent `package-lock.json` ? In the build usually created Release configuration. Does that means that 2 builds for Debug and for Release needed?

pavelsavara replied:
This is in-tree sample of the runtime repo and so it's using npm package as it was built in the `artifacts` folder.
I updated the sample to `npm install` the `@microsoft/dotnet-runtime` from MSBuild/csproj during the build of the sample, so that we would have the same Debug/Release configuration as the C# part. 

The real solution would be to start publishing `@microsoft/dotnet-runtime` npm package into public repository. I don't know if or when we would get there.

campersau commented on src/mono/sample/wasm/browser-webpack/package.json:
```diff
@@ -0,0 +1,21 @@
+  "description": "",
+  "main": "index.js",
+  "scripts": {
+    "build:dotnet": "dnet build /p:TargetOS=Browser /p:TargetArchitecture=wasm /p:Configuration=Debug",
```
```suggestion
    "build:dotnet": "dotnet build /p:TargetOS=Browser /p:TargetArchitecture=wasm /p:Configuration=Debug",
```

pavelsavara committed:
Update src/mono/sample/wasm/browser-webpack/package.json

Co-authored-by: campersau <buchholz.bastian@googlemail.com>

campersau commented on src/mono/sample/wasm/browser-webpack/index.js:
```diff
@@ -0,0 +1,32 @@
+    const element = document.createElement('div');
+
+    const ret = await dotnetMeaning();
+    element.innerHTML = `${ret} as computed on dotnet`;
```
```suggestion
    element.textContent = `${ret} as computed on dotnet`;
```
Avoid XSS!

pavelsavara committed:
Update src/mono/sample/wasm/browser-webpack/index.js

Co-authored-by: campersau <buchholz.bastian@googlemail.com>

pavelsavara committed:
wip

pavelsavara committed:
- put npm package into @microsoft scope and add version info
- fix samples to install the @microsoft/dotnet-runtime during msbuild compilation, so that we don't have to guess build configuration

pavelsavara committed:
- typeScript sample
- fix type definitions

pavelsavara committed:
fix lock files

pavelsavara committed:
fix

pavelsavara committed:
fix

maraf commented on src/mono/sample/wasm/browser-nextjs/Wasm.Browser.NextJs.Sample.csproj:
```diff
@@ -0,0 +1,16 @@
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <WasmMainJSPath>package.json</WasmMainJSPath>
```
I know it is a placeholder in this case. Shouldn't we do something with it here?

pavelsavara replied:
https://github.com/dotnet/runtime/issues/63465

maraf commented on src/mono/sample/wasm/browser-nextjs/components/deepThought.js:
```diff
@@ -0,0 +1,53 @@
+import { useState, useEffect } from 'react'
+import createDotnetRuntime from '@microsoft/dotnet-runtime'
+
+let dotnetRuntime = undefined;
```
React has a way to declare state that is shared between components. Loaded runtime is definitely something that fits in. I'm not sure if we should to keep it simple here or do it the right way.

pavelsavara replied:
I improved the code, but I would like to keep it vanillaJs.

maraf commented on src/mono/sample/wasm/browser-nextjs/Program.cs:
```diff
@@ -0,0 +1,22 @@
+{
+    public class Test
+    {
+        public static void Main(string[] args)
```
The `main` method isn't used, is it?

pavelsavara replied:
The `WasmBuildApp` build target only runs for `Exe`, I removed the other method.

maraf commented on src/mono/sample/wasm/browser-webpack/Program.cs:
```diff
@@ -0,0 +1,22 @@
+{
+    public class Test
+    {
+        public static void Main(string[] args)
```
The `main` method isn't used, is it?

pavelsavara committed:
feedback

maraf reviewed (approved):
LGTM

pavelsavara committed:
- updated js_string_to_mono_string to return MonoStringNull insted of null
- updated dotnet.d.ts banner

pavelsavara committed:
Merge branch 'main' into wasm_sample_nextjs

pavelsavara committed:
no helix

pavelsavara commented:
/azp run runtime-manual

pavelsavara committed:
no CI
